### PR TITLE
chore(gatsby-react-router-scroll): Remove legacy context

### LINKED ISSUE #21231: StrictMode Error: Please update the following components: ScrollContext
I copied this comment from #10179 since it was not really related to that particular error. However this does occur in StrictMode and it would be nice to get this resolved.

It was also posted in this issue #13676 but that has been closed without a resolution.

> I'm getting the following error with `StrictMode`:
> 
> ```
> Warning: Legacy context API has been detected within a strict-mode tree: 
>     in StrictMode (created by App)
>     in App
>     in _default
> 
> Please update the following components: ScrollContext
> 
> Learn more about this warning here:
> https://fb.me/react-strict-mode-warnings
> ```
> 
> It seems that `ScrollContext` is still using legacy context:
> https://github.com/gatsbyjs/gatsby/blob/bddfcbc978da93ae8c13e11622381c8c2a311e4f/packages/gatsby-react-router-scroll/src/ScrollBehaviorContext.js#L14-L16
> 
> Related:
> 
> > Reminder: mere presence of legacy context API in the tree puts everything below it onto a slower code path. It is worth migrating away from.
> > — https://twitter.com/dan_abramov/status/1097866569701621763

### PR DESCRIPTION
**Pull request by herecydev**
## Description

Replace legacy context with React 16.3 Context

## Related Issues

Fixes #21231

### CODE REVIEW

File: packages/gatsby-react-router-scroll/src/ScrollBehaviorContext.js [hunk 1]
```diff
@@ -95,11 +87,20 @@ class ScrollContext extends React.Component {
...
 
   render() {
-    return React.Children.only(this.props.children)
+    return (
+      <ScrollBehaviorContext.Provider
+        value={{
+          registerElement: this.registerElement,
+          unregisterElement: this.unregisterElement,
+          scrollBehavior: this.scrollBehavior,
+        }}
```
Comment (dance2die) [inline packages/gatsby-react-router-scroll/src/ScrollBehaviorContext.js]: Wouldn't this re-render child components whenever this component re-renders?
_because a new object instance passed on each render._

Comment (herecydev) [reply packages/gatsby-react-router-scroll/src/ScrollBehaviorContext.js]: I think you're right. Let me change to use `this`

File: packages/gatsby-react-router-scroll/src/ScrollContainer.js [hunk 2]
```diff
@@ -78,7 +72,12 @@ class ScrollContainer extends React.Component {
   }
 }
 
-ScrollContainer.propTypes = propTypes
-ScrollContainer.contextTypes = contextTypes
+const ScrollContainerConsumer = props => (
+  <ScrollBehaviorContext.Consumer>
+    {context => <ScrollContainer {...props} context={context} />}
+  </ScrollBehaviorContext.Consumer>
+)
```
Comment (blainekasten) [inline packages/gatsby-react-router-scroll/src/ScrollContainer.js]: Shouldn't need this wrapper. You can attach to a single context with the static:

https://reactjs.org/docs/context.html#classcontexttype

Comment (blainekasten) [reply packages/gatsby-react-router-scroll/src/ScrollContainer.js]: Actually, never mind. That wasn't added until 16.6 which would not match our compatibility table for react versions.

Comment (herecydev) [reply packages/gatsby-react-router-scroll/src/ScrollContainer.js]: Really neat, however available from 16.6.0 onwards. Isn't that going to be an issue?

**Comment by blainekasten** [comment]
@herecydev This looks great! We're working on deciding a feature set and timeline for a v3. This is technically a major change as it breaks compatibility for consumers outside of that 16.3 range. Which ironically, this scroll container is used in more than just our codebase.

I think that since this is here to satisfy some warnings, we're going to pause this PR and consider making it part of V3

**Comment by wardpeet** [comment]
Gatsby depends on React 16.4.3 so I published a major of this plugin as it's used internally in gatsby.

Available in gatsby@2.21.15

**Pull request merged**